### PR TITLE
Implement Inverted Waves (- wave levels)

### PR DESCRIPTION
**Pull request by ramen-cat-coder**
Implements the "inverted waves" custom content feature, which allows attacks with negative wave levels that move backwards from the end point to the start point. Currently it does not function for Mini-Waves.

This has also removed the wave level cap, which can be re-implemented if you want to re-implement it.

### CODE REVIEW

**Comment by ramen-cat-coder** [comment]
Due to an extremely strange GitHub glitch I can't exactly edit my fork anymore without this being accepted otherwise my changes will be destroyed.. so if this doesn't work I'll have to restart and make a new PR.

File: battle/attack/AttackSimple.java [hunk 1]
```diff
@@ -156,6 +156,12 @@ public void excuse() {
 			int wid = dire == 1 ? W_E_WID : W_U_WID;
 			float addp = (dire == 1 ? W_E_INI : W_U_INI) + wid / 2f;
 			float p0 = model.getPos() + dire * addp;
+
+			if (proc.WAVE.lv < 0 && model.getDire() != 0) {
```
Comment (MandarinSmell) [inline battle/attack/AttackSimple.java]: In line 160, what's the reason of checking dire here? This means that if dire is 0, inverted wave will work weirdly

File: battle/attack/ContWaveDef.java [hunk 2]
```diff
@@ -51,10 +51,10 @@ public void update() {
 		}
 		if (!activate)
 			return;
-		if (t == (isMini ? W_MINI_TIME : W_TIME)) {
+		if (t == (isMini ? W_MINI_TIME - 1 : W_TIME)) {
```
Comment (MandarinSmell) [inline battle/attack/ContWaveDef.java:54]: I want to know the reason of -1 in line 54. If this pull request is about adding inverted wave, you shouldn't touch other mechanisms

File: util/lang/Editors.java [hunk 3]
```diff
@@ -306,7 +306,7 @@ else if (t.mult == 0)
 
 		map().put("WAVE", new EditControl<>(Proc.WAVE.class, (t) -> {
 			t.prob = MathUtil.clip(t.prob, 0, 100);
-			t.lv = MathUtil.clip(t.lv, 1, 20);
+			t.lv = t.lv != 0 ? t.lv : 1;
```
Comment (MandarinSmell) [inline util/lang/Editors.java:309]: In line 309, `MathUtil.clip` is for clipping the value, basically this means `Math.max(1, Math.min(t.lv, 20))`. Due to this change, clipping will be removed, and users will be able to put level something like 100. Please clip the level

File: util/lang/Editors.java [hunk 4]
```diff
@@ -586,7 +586,7 @@ else if (t.mult == 0)
 				t.lv = 0;
 				t.multi = 0;
 			} else {
-				t.lv = MathUtil.clip(t.lv, 1, 20);
+				t.lv = MathUtil.clip(t.lv, 1, 20);;
```
Comment (MandarinSmell) [inline util/lang/Editors.java:589]: Redundant `;` letter found in line 589

**Review by MandarinSmell — CHANGES_REQUESTED** [review]
I have commented questions and pointed out the problems/warnings. Please change the codes. Also, instead of using negative level, it will be better to add new flag in `WAVE`, `MINIWAVE` class. This is because someday other developer or Ponos may want to add wave going backwards (starting from entity's position, not end position of final wave, which is about current pull request). Add `boolean inverted` into the class, and using this flag will be better imho

File: battle/attack/AttackSimple.java (hunk 1 above)
Comment (ramen-cat-coder) [reply battle/attack/AttackSimple.java]: It was part of my initial fixes when I ran into bugs with wave positioning. Now it is redundant and will be removed.

File: battle/attack/ContWaveDef.java (hunk 2 above)
Comment (ramen-cat-coder) [reply battle/attack/ContWaveDef.java:54]: It was shown as a conflict in the GitHub summary.

**Comment by ramen-cat-coder** [comment]
I think I'll redo the feature with a flag, so it behaves properly and I can get rid of the GitHub error. it was not too much of a problem to implement in the first place.

**Pull request closed without merging**